File: src/js-interpreter/src/ast.js
/**
 * AST classes for Smalltalk interpreter
 */

class Node {
  constructor() {}
}

class Program extends Node {
  constructor(statements) {
    super();
    this.statements = statements || [];
  }

  evaluate(context) {
    let result;
    for (const statement of this.statements) {
      result = statement.evaluate(context);
    }
    return result;
  }
}

class Literal extends Node {
  constructor(value) {
    super();
    this.value = value;
  }

  evaluate(context) {
    return this.value;
  }
}

class Variable extends Node {
  constructor(name) {
    super();
    this.name = name;
  }

  evaluate(context) {
    return context.lookup(this.name);
  }
}

class Assignment extends Node {
  constructor(variable, expression) {
    super();
    this.variable = variable;
    this.expression = expression;
  }

  evaluate(context) {
    const value = this.expression.evaluate(context);
    context.assign(this.variable.name, value);
    return value;
  }
}

class MessageSend extends Node {
  constructor(receiver, selector, args) {
    super();
    this.receiver = receiver;
    this.selector = selector;
    this.args = args || [];
  }

  evaluate(context) {
    const receiver = this.receiver.evaluate(context);
    const args = this.args.map(arg => arg.evaluate(context));
    return receiver.sendMessage(this.selector, args, context);
  }
}

class Block extends Node {
  constructor(parameters, statements) {
    super();
    this.parameters = parameters || [];
    this.statements = statements || [];
  }

  evaluate(context) {
    // Create a Block object that captures the current context
    return context.createBlock(this.parameters, this.statements, context);
  }
}

module.exports = {
  Node,
  Program,
  Literal,
  Variable,
  Assignment,
  MessageSend,
  Block
};

File: src/interpreter/ast/ast.go
package ast

import (
	"smalltalklsp/interpreter/core"
)

// Node is the interface for all AST nodes
type Node interface {
	// Accept accepts a visitor
	Accept(visitor Visitor) interface{}
}

// Visitor is the interface for visitors
type Visitor interface {
	// VisitMethodNode visits a method node
	VisitMethodNode(node *MethodNode) interface{}

	// VisitReturnNode visits a return node
	VisitReturnNode(node *ReturnNode) interface{}

	// VisitSelfNode visits a self node
	VisitSelfNode(node *SelfNode) interface{}

	// VisitLiteralNode visits a literal node
	VisitLiteralNode(node *LiteralNode) interface{}

	// VisitVariableNode visits a variable node
	VisitVariableNode(node *VariableNode) interface{}

	// VisitAssignmentNode visits an assignment node
	VisitAssignmentNode(node *AssignmentNode) interface{}

	// VisitMessageSendNode visits a message send node
	VisitMessageSendNode(node *MessageSendNode) interface{}

	// VisitBlockNode visits a block node
	VisitBlockNode(node *BlockNode) interface{}
}

// MethodNode represents a method definition
type MethodNode struct {
	// Selector is the method selector
	Selector string

	// Parameters are the method parameters
	Parameters []string

	// Temporaries are the method temporaries
	Temporaries []string

	// Body is the method body
	Body Node

	// Class is the method class
	Class *core.Object
}

// Accept implements the Node interface
func (n *MethodNode) Accept(visitor Visitor) interface{} {
	return visitor.VisitMethodNode(n)
}

// ReturnNode represents a return statement
type ReturnNode struct {
	// Expression is the expression to return
	Expression Node
}

// Accept implements the Node interface
func (n *ReturnNode) Accept(visitor Visitor) interface{} {
	return visitor.VisitReturnNode(n)
}

// SelfNode represents the self reference
type SelfNode struct{}

// Accept implements the Node interface
func (n *SelfNode) Accept(visitor Visitor) interface{} {
	return visitor.VisitSelfNode(n)
}

// LiteralNode represents a literal value
type LiteralNode struct {
	// Value is the literal value
	Value *core.Object
}

// Accept implements the Node interface
func (n *LiteralNode) Accept(visitor Visitor) interface{} {
	return visitor.VisitLiteralNode(n)
}

// VariableNode represents a variable reference
type VariableNode struct {
	// Name is the variable name
	Name string
}

// Accept implements the Node interface
func (n *VariableNode) Accept(visitor Visitor) interface{} {
	return visitor.VisitVariableNode(n)
}

// AssignmentNode represents an assignment
type AssignmentNode struct {
	// Variable is the variable to assign to
	Variable string

	// Expression is the expression to assign
	Expression Node
}

// Accept implements the Node interface
func (n *AssignmentNode) Accept(visitor Visitor) interface{} {
	return visitor.VisitAssignmentNode(n)
}

// MessageSendNode represents a message send
type MessageSendNode struct {
	// Receiver is the message receiver
	Receiver Node

	// Selector is the message selector
	Selector string

	// Arguments are the message arguments
	Arguments []Node
}

// Accept implements the Node interface
func (n *MessageSendNode) Accept(visitor Visitor) interface{} {
	return visitor.VisitMessageSendNode(n)
}

// BlockNode represents a block
type BlockNode struct {
	// Parameters are the block parameters
	Parameters []string

	// Temporaries are the block temporaries
	Temporaries []string

	// Body is the block body
	Body Node
}

// Accept implements the Node interface
func (n *BlockNode) Accept(visitor Visitor) interface{} {
	return visitor.VisitBlockNode(n)
}
